### PR TITLE
Feat: add update account setting page

### PR DESCRIPTION
**Pull request by SeijaHuang**
https://beequant.atlassian.net/jira/software/projects/CP/boards/1?selectedIssue=CP-113
---

## What happens here?
Add app setting page:
- Implemented language change
- Implemented theme change Add internationalization

Add next-intl library to achieve internationalization routing
- Modify middleware with next-intl middleware
- Modify next navigations with next-intl shared routing, including redirect, useRouter etc. Resolve CP-113

Screenshot:
1. Login Page
10001724203600_ pic
10011724203612_ pic
10031724203717_ pic
10021724203700_ pic

2. Rigester Page
10041724203754_ pic
10051724203755_ pic
10061724203781_ pic
10071724203781_ pic

4. Dashboard
10091724203866_ pic
10081724203851_ pic

6. App Setting Page
10101724204036_ pic
10111724204055_ pic

7. Account Setting Page
10121724204158_ pic
10131724204234_ pic

8. Register Succeed Page
10151724205221_ pic
10141724205209_ pic

## How do I know this is working?
in this branch,  you can use /en/pathname or /zh-cn/pathname to see all changes

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,16 +1,10 @@
 # BeeQuant Turbo
 
-## How do I run this
-
-### Preping local database
+## What is this?
```
Comment (NorrisWu0) [inline README.md]: What's going on here?

File: apps/client/package.json [hunk 2]
```diff
@@ -2,7 +2,7 @@
   "name": "@beequant/client",
   "version": "0.1.0",
   "scripts": {
-    "dev": "concurrently 'yarn:dev-*' --restart-tries 3 --restart-after 3000",
+    "dev": "yarn dev-codegen & yarn dev-next",
```
Comment (NorrisWu0) [inline apps/client/package.json]: yeah, you need to fix these kind of changes before I can do reviews

**Review by NorrisWu0 — CHANGES_REQUESTED** [review]

File: apps/client/package.json (hunk 2 above)
Comment (SeijaHuang) [reply apps/client/package.json]: reset with dev branch

File: README.md (hunk 1 above)
Comment (SeijaHuang) [reply README.md]: reset with dev branch

File: apps/client/src/module/appsetting/app-setting-page.tsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    localesSwitchRules,
+    loading,
+  } = useLanguageSwitcher();
+  const {
+    currentTheme,
+    setCurrentTheme,
+    switchThemeToOption,
+    switchOptionToTheme,
+    isTheme,
+  } = useThemeSwitcher();
```
Comment (NorrisWu0) [inline apps/client/src/module/appsetting/app-setting-page.tsx]: there's already a existing theme switcher, check under `module/system`

File: apps/client/src/module/appsetting/language-switcher.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  };
+  const optionalLocales = ["en", "zh-cn"];
+
+  const switchLanguageValue = (
+    lang: keyof typeof localesSwitchRules,
+  ): string => {
+    return localesSwitchRules[lang];
+  };
+
+  const isLocale = (value: string): value is locales => {
```
Comment (NorrisWu0) [inline apps/client/src/module/appsetting/language-switcher.ts]: ~so this looks like a util function? why is this here?~

we provides the options to switch languages yeah? why do you need to check if a value is a locale?

and why do you need `value is locales`?

File: apps/client/src/module/appsetting/language-switcher.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
+"use client";
+
+import { getNextLocale } from "@src/actions";
+import { useEffect, useState } from "react";
+
+type locales = "en" | "zh-cn";
```
Comment (NorrisWu0) [inline apps/client/src/module/appsetting/language-switcher.ts]: You know what to do here

File: apps/client/src/module/appsetting/language-switcher.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  useEffect(() => {
+    getNextLocale().then((locale) => {
+      if (isLocale(locale)) {
+        setCurrentLocale(locale);
+        setLoading(false);
+      }
+    });
+  }, []);
+
+  // Locales switch function and rules
```
Comment (NorrisWu0) [inline apps/client/src/module/appsetting/language-switcher.ts]: ummmm, why do you need the switch functions and rules? we know what the locales are yeah?

File: apps/client/src/module/appsetting/app-setting-page.tsx [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+  const { setTheme } = useTheme();
+
+  const options = {
+    language: ["English", "中文"],
+    theme: [t("Shared.darkTheme"), t("Shared.lightTheme")],
+  };
+
+  const { control } = useForm<FormValues>({
```
Comment (NorrisWu0) [inline apps/client/src/module/appsetting/app-setting-page.tsx]: first up, why do you need to use `react-hook-form` here?

second, it doesn't seem to be used for anything?

File: apps/client/src/module/common/select.tsx [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  const { field } = useController({ name, control });
+
+  return (
+    <label className="grid grid-cols-[80px_1fr] items-center gap-x-[10px] space-y-1 sm:grid-cols-[120px_1fr] sm:gap-x-5">
+      <span>{label}</span>
+      <select
+        className="focus:shadow-accent-400 dark:focus:shadow-accent-400 shadow-primary-100 dark:shadow-primary-600 shadow-inset flex-1 bg-transparent px-2 py-2 text-sm transition-[box-shadow] duration-300 placeholder:text-sm focus:outline-none"
+        {...field}
+        {...props}
+        onChange={(e) => onChangeHandler(e)}
```
Comment (NorrisWu0) [inline apps/client/src/module/common/select.tsx]: is this working? if you don't want react-hook-form to control this input, then you probably don't need the `useController` hook yeah?

File: apps/client/src/module/common/icon.tsx [hunk 9]
```diff
@@ -82,9 +82,7 @@ type LinearIconProps = { icon: LinearIcons; className?: string };
 export function LinearIcon({ icon, className }: LinearIconProps) {
   return (
     <span
-      className={twMerge(
-        clsx(`lnr lnr-${icon} text-primary-200 text-base`, className),
-      )}
+      className={twMerge(clsx(`lnr lnr-${icon} text-primary-200`, className))}
```
Comment (NorrisWu0) [inline apps/client/src/module/common/icon.tsx]: oh, what's wrong with `text-base`?

File: apps/client/src/module/common/controlled-input.tsx [hunk 10]
```diff
@@ -50,7 +50,7 @@ export function ControlledTextInput<TFieldValues extends object>({
 
   return (
     <label className={inputVariants({ direction })}>
-      <span>
+      <span className="text-[13px]">
```
Comment (NorrisWu0) [inline apps/client/src/module/common/controlled-input.tsx]: we got `text-base` yeah? do we need arbitrary value here?

File: apps/client/src/module/dashboard/layout/sidebar-menu.tsx [hunk 11]
```diff
@@ -45,8 +45,8 @@ function SidebarMenuItem(props: SidebarMenuItem) {
         <span
```
Comment (NorrisWu0) [inline apps/client/src/module/dashboard/layout/sidebar-menu.tsx]: I'm gonna assume these style changes are fine

File: apps/client/src/module/dashboard/account/setting/setting-form.tsx [hunk 12]
```diff
@@ -3,40 +3,37 @@ import { zodResolver } from "@hookform/resolvers/zod";
...
-      .min(4, { message: DisplayErrorMsgs.MinLength })
-      .max(15, { message: DisplayErrorMsgs.MaxLength })
-      .regex(displayNamePatten, {
-        message: DisplayErrorMsgs.Invalid,
-      }),
-    z
+export function AccountSettingForm() {
+  const t = useTranslations();
+
+  const formSchema = z.object({
```
Comment (NorrisWu0) [inline apps/client/src/module/dashboard/account/setting/setting-form.tsx:12]: assuming this schema is good

File: apps/client/src/module/appsetting/theme-switcher.ts [hunk 13]
```diff
@@ -0,0 +1,53 @@
+"use client";
```
Comment (NorrisWu0) [inline apps/client/src/module/appsetting/theme-switcher.ts]: not gonna comment on this, since we already got one

File: apps/client/src/module/appsetting/language-switcher.ts [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    中文: "zh-cn",
+    en: "English",
+    "zh-cn": "中文",
+  };
+  const optionalLocales = ["en", "zh-cn"];
+
+  const switchLanguageValue = (
+    lang: keyof typeof localesSwitchRules,
+  ): string => {
+    return localesSwitchRules[lang];
```
Comment (NorrisWu0) [inline apps/client/src/module/appsetting/language-switcher.ts]: so what does this function do?

File: apps/client/src/module/appsetting/app-setting-page.tsx [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    <div className="dark:bg-primary-900 bg-primary-50 shadow-settingPage dark:shadow-settingPage flex flex-col rounded-lg p-5">
+      <div className="space-y-1">
+        <h1 className="text-base font-bold">{t("AppSettingPage.title")}</h1>
+        <h5 className="text-primary-700 dark:text-primary-100 text-sm opacity-70">
+          {t("AppSettingPage.subtitle")}
+        </h5>
+      </div>
+      {!loading && (
+        <form className="my-[30px] flex flex-col gap-y-5">
+          <Select
```
Comment (NorrisWu0) [inline apps/client/src/module/appsetting/app-setting-page.tsx:43]: again, it doesn't seems like you need a controlled select here.

File: apps/client/src/middleware.ts [hunk 16]
```diff
@@ -1,16 +1,27 @@
 import { NextRequest, NextResponse } from "next/server";
```
Comment (NorrisWu0) [inline apps/client/src/middleware.ts:1]: assuming this is all good

File: apps/client/src/i18n.ts [hunk 17]
```diff
@@ -0,0 +1,14 @@
+import { notFound } from "next/navigation";
```
Comment (NorrisWu0) [inline apps/client/src/i18n.ts:1]: assuming this is all good

**Review by NorrisWu0 — CHANGES_REQUESTED** [review]
okay, well this is a lot, A LOT of changes. going to assume the files I've haven't look at are good (mostly files under app directory and static files),

File: apps/client/src/module/appsetting/language-switcher.ts (hunk 4 above)
Comment (SeijaHuang) [reply apps/client/src/module/appsetting/language-switcher.ts]: Changed using as

File: apps/client/src/module/appsetting/language-switcher.ts (hunk 5 above)
Comment (SeijaHuang) [reply apps/client/src/module/appsetting/language-switcher.ts]: yes, import form the config file, the only place stored optional locales

File: apps/client/src/module/appsetting/language-switcher.ts (hunk 6 above)
Comment (SeijaHuang) [reply apps/client/src/module/appsetting/language-switcher.ts]: changed the data structure as you mentored

File: apps/client/src/module/appsetting/app-setting-page.tsx (hunk 7 above)
Comment (SeijaHuang) [reply apps/client/src/module/appsetting/app-setting-page.tsx]: yes, there is no need here

File: apps/client/src/module/common/select.tsx (hunk 8 above)
Comment (SeijaHuang) [reply apps/client/src/module/common/select.tsx]: actually in the previous file, there is a useController hook

File: apps/client/src/module/common/icon.tsx (hunk 9 above)
Comment (SeijaHuang) [reply apps/client/src/module/common/icon.tsx]: reset all these changes

File: apps/client/src/module/common/controlled-input.tsx (hunk 10 above)
Comment (SeijaHuang) [reply apps/client/src/module/common/controlled-input.tsx]: reset all these changes

File: apps/client/src/module/appsetting/language-switcher.ts (hunk 14 above)
Comment (SeijaHuang) [reply apps/client/src/module/appsetting/language-switcher.ts]: changed the data structure as you mentored

File: apps/client/src/module/appsetting/app-setting-page.tsx (hunk 15 above)
Comment (SeijaHuang) [reply apps/client/src/module/appsetting/app-setting-page.tsx:43]: yeah， already delete react form hook

File: apps/client/next.config.mjs [hunk 18]
```diff
@@ -1,6 +1,10 @@
+import createNextIntlPlugin from "next-intl/plugin";
+
+const withNextIntl = createNextIntlPlugin();
+
 /** @type {import('next').NextConfig} */
```
Comment (Chris-Wang) [inline apps/client/next.config.mjs:5]: what's this comment for?

**Review by NorrisWu0 — DISMISSED** [review]

File: apps/client/src/actions/get-next-locale.ts [hunk 19]
```diff
@@ -0,0 +1,13 @@
...
+import { defaultLocale } from "../configs/configs";
+import { cookies } from "next/headers";
+
+export async function getNextLocale() {
+  const cookieStore = cookies();
+  const result = await cookieStore.get("NEXT_LOCALE")?.value;
+  if (!result) {
+    return defaultLocale;
+  }
+  const nextLocale = result;
```
Comment (Chris-Wang) [inline apps/client/src/actions/get-next-locale.ts]: why not directly return result?

File: apps/client/src/app/[locale]/(auth)/register/page.tsx [hunk 20]
```diff
@@ -1,4 +1,4 @@
-import { AuthRoute, RegisterForm } from "../../../module/auth";
+import { AuthRoute, RegisterForm } from "../../../../module/auth";
```
Comment (Chris-Wang) [inline apps/client/src/app/[locale]/(auth)/register/page.tsx]: don't we have alias of module, we probably think of using @ for import

File: apps/client/src/utils/user-help-messgage.ts [hunk 21]
```diff
@@ -1,5 +1,3 @@
 export enum UserHelpMsgs {
   DisplayNameSettingPage = "Display is optional: it should between 4 to 15 characters, only letters, numbers, dashes and underscore are allowed!",
```
Comment (Chris-Wang) [inline apps/client/src/utils/user-help-messgage.ts]: is `DisplayNameSettingPage` still being used?

File: apps/client/src/module/system/system.ts [hunk 22]
```diff
@@ -0,0 +1,23 @@
+"use client";
+
+import { Locales } from "@src/configs/configs";
+import { create } from "zustand";
+
+export type Theme = "dark" | "light" | "system";
```
Comment (Chris-Wang) [inline apps/client/src/module/system/system.ts]: `system` for ?

File: apps/client/src/module/system/theme-switcher.tsx [hunk 23]
```diff
@@ -110,7 +82,12 @@ export function useTheme() {
     };
   }, [setTheme]);
 
-  return { theme, setTheme };
+  function themeSwitcher({ e }: { e: React.ChangeEvent<HTMLSelectElement> }) {
```
Comment (Chris-Wang) [inline apps/client/src/module/system/theme-switcher.tsx:100]: I like this refactoring

File: apps/client/src/middleware.ts [hunk 24]
```diff
@@ -1,16 +1,27 @@
...
+
+const handleI18Routing = createMiddleware({
+  locales,
+  defaultLocale,
+});
 
 export async function middleware(request: NextRequest) {
   const token = request.cookies.get("token");
+  const pathname = request.nextUrl.pathname;
+  const pagesWithoutToken =
```
Comment (Chris-Wang) [inline apps/client/src/middleware.ts]: could we have an array to store those page names for better maintainability

**Review by Chris-Wang — CHANGES_REQUESTED** [review]
better to split PR into smaller ones the next time, as we discussed in the standup

File: apps/client/src/actions/get-next-locale.ts (hunk 19 above)
Comment (SeijaHuang) [reply apps/client/src/actions/get-next-locale.ts]: Because the result of NEXT_LOCALE maybe undefined(user never visited our page)
image type undefined is not assignable to the locales('en' | 'zh-cn'), so if result is undefined it needs to return defaultLocale, which is 'en'

File: apps/client/src/module/system/system.ts (hunk 22 above)
Comment (SeijaHuang) [reply apps/client/src/module/system/system.ts]: system setting includes language and theme setting, there two global states are system scope, so I gave the name system for them

File: apps/client/src/actions/get-next-locale.ts (hunk 19 above)
Comment (Chris-Wang) [reply apps/client/src/actions/get-next-locale.ts]: at line 11, the result shouldn't be falsy, so what's the nextLocale for?

File: apps/client/src/utils/user-help-messgage.ts (hunk 21 above)
Comment (SeijaHuang) [reply apps/client/src/utils/user-help-messgage.ts]: No, I've deleted it already.

File: apps/client/src/app/[locale]/(auth)/register/page.tsx (hunk 20 above)
Comment (SeijaHuang) [reply apps/client/src/app/[locale]/(auth)/register/page.tsx]: refactored

File: apps/client/src/module/system/system.ts (hunk 22 above)
Comment (Chris-Wang) [reply apps/client/src/module/system/system.ts]: when could the `Theme` be `system`? 

I assume `Theme` represent colour theme of the system here, do we make it like color theme or language theme here?

File: apps/client/src/actions/get-next-locale.ts (hunk 19 above)
Comment (SeijaHuang) [reply apps/client/src/actions/get-next-locale.ts]: The NEXT_LOCALE is set by next-intl. Once user have visited or changed language the language setting will be stored in the cookie.  In this case,  by getting the preferred language or default language from the cookie, the setting page select will render the responding language setting.

File: apps/client/src/utils/user-help-messgage.ts (hunk 21 above)
Comment (Chris-Wang) [reply apps/client/src/utils/user-help-messgage.ts]: if UserHelpMsgs is unused, why not remove it?

File: apps/client/next.config.mjs (hunk 18 above)
Comment (SeijaHuang) [reply apps/client/next.config.mjs:5]: I would delete it!

File: apps/client/src/actions/get-next-locale.ts (hunk 19 above)
Comment (Chris-Wang) [reply apps/client/src/actions/get-next-locale.ts]: ... 

shouldn't we remove line 11 and 12, and just make it
` return result`?

File: apps/client/next.config.mjs (hunk 18 above)
Comment (SeijaHuang) [reply apps/client/next.config.mjs:5]: This is form next doc, which is used to add some type checking in config fiels.
image

File: apps/client/src/module/system/system.ts (hunk 22 above)
Comment (SeijaHuang) [reply apps/client/src/module/system/system.ts]: delete this system.ts, put language and theme settings back to their responding files

File: apps/client/src/middleware.ts (hunk 24 above)
Comment (SeijaHuang) [reply apps/client/src/middleware.ts]: Sure!

**Review by Chris-Wang — APPROVED** [review]
LGTM
